fix: Prevent discarding editor WebView state during orientation change

dcalhoun opened the pull request:
The GutenbergKit editor relies upon a WebView, which discarded all its
state—including content edits—when the device orientation changed.

`EditPostActivity` is currently also utilized by Aztec and Gutenberg
Mobile, so there could be negative side effects to those fragments with
this change.

This also removes an unnecessary `configChanges` value from the 
experimental features fragment. 

Fix https://github.com/wordpress-mobile/GutenbergKit/issues/71.

## To Test:
<!-- Test instructions per dependency update: https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/docs/test_instructions_per_dependency_update.md -->
1. Enable the experimental editor setting.
1. Launch the editor.
1. Modify the title and/or content.
1. Rotate the device.
1. Verify the title/content state is preserved.

## Regression Notes
1. Potential unintended areas of impact
    Regressions in the Aztec and Gutenberg Mobile editors.
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Manually tested the editors.
3. What automated tests I added (or what prevented me from doing so)
    Deemed unnecessary for the experimental editor.

## PR Submission Checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Testing Checklist (strike-out the not-applying and unnecessary ones):

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

dcalhoun committed:
fix: Prevent discarding editor WebView state during orientation change

The GutenbergKit editor relies upon a WebView, which discarded all its
state--including content edits--when the device orientation changed.
`EditPostActivity` is currently also utilized by Aztec and Gutenberg
Mobile, so there could be negative side effects to those fragments with
this change.

dcalhoun committed:
fix: Experimental features activity relies upon OS-applied configuration

There is no explicit need for this configuration, it was mistakenly
copied during the initial creation.

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

dcalhoun committed:
docs: Add release note

dcalhoun commented:
@nbradbury Currently, all three editors—GutenbergKit, Gutenberg Mobile, and Aztec—share the same `EditPostActivity.kt`. I'd like your honest opinion on the proposed changes—namely, how likely it is they might create problems for the Aztec and Gutenberg Mobile editors? 

I tested all three editors and did not experience any issues. However, my lack of Android experience leaves me wondering if this type of change could be problematic—e.g., performance, memory leaks, crashes. 

In some ways, refactoring to split up `EditPostActivity.kt` is an attractive thought. It would negate my concerns and simplify that file significantly. However, it'd be a lot of work and would push complexity upwards into more areas of the app— spreading editor-type conditionals in various locations, all the numerous locations the editor is launched.

nbradbury commented:
Overriding `configChanges` like this is often seen as [a bandaid](https://stackoverflow.com/a/7990543/1673548) but I think we're okay in this case, especially since it solves the content loss issue and works fine with the other editors.

Note, though, that we still have content loss when the app is killed. To reproduce this:

* Edit a post
* Kill the app
* Return to the app
* Changes are lost

I imagine we would see the same problem in other situations where the app is killed, including changing dark mode or changing language. This happens regardless of `configChanges`, though, so it's not specific to this PR.

Also, a cursory look didn't see any potential memory leaks with this change.

nbradbury reviewed (approved):
Merge when ready! :shipit:

dcalhoun commented:
> Note, though, that we still have content loss when the app is killed.

@nbradbury content loss when quitting the app appears resolved by https://github.com/wordpress-mobile/WordPress-Android/pull/21663, at least if one waits long enough for an auto-save event to occur. 

> I imagine we would see the same problem in other situations where the app is killed, including changing dark mode or changing language.

Oof, I didn't consider dark mode changes. Is there a reasonable way for us to prevent this loss?

nbradbury commented:
> I didn't consider dark mode changes. Is there a reasonable way for us to prevent this loss?

You can add `uiMode` to `configChanges` but I believe that may prevent the activity from reflecting the change to/from dark mode.

As an aside, [this post](https://developer.android.com/quick-guides/content/manage-webview-state) recommends using `webView.invalidate()` when the configuration changes.

dcalhoun commented:
@nbradbury thank you for sharing the additional configuration and helpful post. I'll experiment with those options to determine how it impacts the three editors. 

Additionally, I may have identified an approach for retaining the editor content modifications that does not involve modifying `configChanges`. The approach is dependent upon https://github.com/wordpress-mobile/WordPress-Android/pull/21663.

dcalhoun committed:
fix: Increase robustness of managing editor configuration changes

Handle additional configurations and invalidate the editor WebView to
follow best practices.

https://developer.android.com/quick-guides/content/manage-webview-state

dcalhoun commented:
@nbradbury I opted to stick with the `configChanges` approach, as it retains scroll position in addition to content changes. I expanded the implementation to mirror the [guide post](https://developer.android.com/quick-guides/content/manage-webview-state) you previously shared. The expanded approach handles additional events I had not considered before, like resizing a window: 

<details><summary>Resize window screen recording</summary>
<p>


https://github.com/user-attachments/assets/3017b6cd-e987-4664-bb5d-51c3c90a9593



</p>
</details> 

I did not add `uiMode` to `configChanges`, as it does disrupt the Aztec and Gutenberg Mobile's ability to appropriately update its UI. 

I think this is the best approach for now. Hopefully we do not encounter regressions in the editor.
